### PR TITLE
non-exp CAPZ types support

### PR DESCRIPTION
**Pull request by whites11**
Towards: https://github.com/giantswarm/giantswarm/issues/22430

azure release 17.1.0 uses non experimental CAPZ types and this requires  a lot of changes to this admission controller.

- we need to use the newest CAPZ dependency (and latest CAPI as a consequence)
- we need to take into account both API versions while validating and defaulting CRs.

This huge PR does just that.

I understand this is a nightmare to review, not sure there are any alternatives to this.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,120 +1,60 @@
 module github.com/giantswarm/azure-admission-controller
 
-go 1.17
+go 1.16
```
Comment (fiunchinho) [inline go.mod:3]: from 1.17 to 1.16?

File: pkg/azurecluster/mutate_update_test.go [hunk 2]
```diff
@@ -25,30 +24,30 @@ func TestAzureClusterUpdateMutate(t *testing.T) {
...
-			patches: []mutator.PatchOperation{
-				{
-					Operation: "add",
-					Path:      "/metadata/labels/cluster-operator.giantswarm.io~1version",
-					Value:     "0.23.11",
-				},
-			},
-			errorMatcher: nil,
-		},
+		//{
```
Comment (fiunchinho) [inline pkg/azurecluster/mutate_update_test.go:27]: don't they make sense anymore?

File: pkg/azurecluster/mutate_create_test.go [hunk 3]
```diff
@@ -26,59 +24,59 @@ func TestAzureClusterCreateMutate(t *testing.T) {
...
-			patches: []mutator.PatchOperation{
-				{
-					Operation: "add",
-					Path:      "/metadata/labels/azure-operator.giantswarm.io~1version",
-					Value:     "5.0.0",
-				},
-			},
-			errorMatcher: nil,
-		},
+		//{
```
Comment (whites11) [inline pkg/azurecluster/mutate_create_test.go:27]: These tests fail in weird ways and I'm not able to make them green.
Sorry don't have time to invest in this right now.

File: pkg/azurecluster/mutate_update_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (fiunchinho) [reply pkg/azurecluster/mutate_update_test.go:27]: I'd delete them in that case.

File: go.mod [hunk 4]
```diff
@@ -1,120 +1,60 @@
...
-	k8s.io/klog/v2 v2.0.0 // indirect
-	k8s.io/kube-openapi v0.0.0-20200410145947-61e04a5be9a6 // indirect
-	k8s.io/utils v0.0.0-20200912215256-4140de9c8800 // indirect
-	sigs.k8s.io/structured-merge-diff/v3 v3.0.1 // indirect
+	k8s.io/api v0.24.1
+	k8s.io/apiextensions-apiserver v0.24.1
+	k8s.io/apimachinery v0.24.1
+	k8s.io/client-go v0.24.1
+	sigs.k8s.io/cluster-api v1.1.4
+	sigs.k8s.io/cluster-api-provider-azure v1.3.2
```
[the remark this answers is not in the document]
Comment (whites11) [reply go.mod:33]: that doesn work (go mod reverts to 1.3.2)
I think the reason is apiextension using 1.3.2 https://github.com/giantswarm/apiextensions/blob/master/go.mod#L17

File: pkg/azurecluster/mutate_update_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (whites11) [reply pkg/azurecluster/mutate_update_test.go:27]: The reason why I'm not deleting them is I don't want to lose the idea behind them. They are still valid tests, just to be implemented.
But I'll leave the decision to you about this. LMK

File: go.mod (hunk 4 above)
[the remark this answers is not in the document]
Comment (fiunchinho) [reply go.mod:33]: You can use `replace` to force whatever version you want. I'd try 1.0.2

[the remark this answers is not in the document]
Comment (whites11) [reply go.mod:33]: I tried that, both versions got installed defusing the whole point I guess

File: go.mod (hunk 1 above)
Comment (fiunchinho) [reply go.mod:3]: you mean during CI? I've seen this, but normally it only happens the first time locally, then I never see it again after running `go mod tidy -go=1.16 && go mod tidy -go=1.17`.

File: pkg/azurecluster/mutate_update_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (fiunchinho) [reply pkg/azurecluster/mutate_update_test.go:27]: ok, let's keep them then.

File: go.mod (hunk 4 above)
[the remark this answers is not in the document]
Comment (fiunchinho) [reply go.mod:33]: mm... I thought you could only have one major version at the same time. If it's possible to have several versions of the same major version then I don't understand how any of this works.

File: go.mod (hunk 1 above)
Comment (whites11) [reply go.mod:3]: yeah it made the CI fail. I can try again

Comment (fiunchinho) [reply go.mod:3]: We are using 1.18 in circleci, maybe worth trying that? I mean upgrading locally, running go mod tidy locally and pushing that.

Comment (whites11) [reply go.mod]: yeah, that worked :bow:

**Pull request merged**